### PR TITLE
PE-3697: large downloads

### PR DESCRIPTION
**Pull request by thiagocarvalhodev**

### CODE REVIEW

**Review by karlprieb — DISMISSED** [review]

**Review by matibat — DISMISSED** [review]

**Review by matibat — APPROVED** [review]

File: lib/components/upload_form.dart [hunk 1]
```diff
@@ -39,14 +43,15 @@ Future<void> promptToUpload(
 }) async {
   final selectedFiles = <UploadFile>[];
   final io = ArDriveIO();
+  IOFolder? ioFolder;
```
Comment (matibat) [inline lib/components/upload_form.dart:46]: I'd prefer to name this parameter differently (e.g. `maybeIoFolder`) so we can have the declaration of `ioFolder` be final.

```dart
final IOFolder ioFolder = maybeIoFolder ?? await io.pickFolder();
```

File: lib/components/upload_form.dart [hunk 2]
```diff
@@ -137,6 +144,11 @@ class UploadForm extends StatefulWidget {
 class _UploadFormState extends State<UploadForm> {
   final _scrollController = ScrollController();
 
+  @override
+  initState() {
+    super.initState();
+  }
+
```
Comment (matibat) [inline lib/components/upload_form.dart:151]: Unecessary
```suggestion
```

File: lib/components/upload_form.dart [hunk 3]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
       );
+
+  Widget _uploadUsingNewUploader({
+    required UploadInProgressUsingNewUploader state,
+  }) {
+    final progress = state.progress;
+    return ArDriveStandardModal(
+      width: kLargeDialogWidth,
+      title:
+          '${appLocalizationsOf(context).uploadingNFiles(state.progress.getNumberOfItems())} ${(state.totalProgress * 100).toStringAsFixed(2)}%',
```
Comment (matibat) [inline lib/components/upload_form.dart:773]: Let's break this line into many final declarations.

File: lib/components/upload_form.dart [hunk 4]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+                padding: const EdgeInsets.all(8),
+                child: Scrollbar(
+                  child: ListView.builder(
+                    shrinkWrap: true,
+                    itemCount: progress.task.length,
+                    itemBuilder: (BuildContext context, int index) {
+                      final task = progress.task[index];
+
+                      String progressText;
+                      String status = '';
```
Comment (matibat) [inline lib/components/upload_form.dart:791]: Make them be late

File: lib/components/upload_form.dart [hunk 5]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+            child: ConstrainedBox(
+              constraints: const BoxConstraints(maxHeight: 256 * 1.5),
+              child: Container(
+                padding: const EdgeInsets.all(8),
+                child: Scrollbar(
+                  child: ListView.builder(
+                    shrinkWrap: true,
+                    itemCount: progress.task.length,
+                    itemBuilder: (BuildContext context, int index) {
+                      final task = progress.task[index];
```
Comment (matibat) [inline lib/components/upload_form.dart:788]: Here `task` seems to be `dynamic`, right? Let's give it a type.

File: lib/components/upload_form.dart [hunk 6]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+                          break;
+                        case UploadStatus.preparationDone:
+                          status = 'Preparation done';
+                          break;
+                      }
+
+                      if (task.isProgressAvailable) {
+                        if (task.uploadItem != null) {
+                          progressText =
+                              '${filesize(((task.uploadItem!.size) * task.progress).ceil())}/${filesize(task.uploadItem!.size)}';
```
Comment (matibat) [inline lib/components/upload_form.dart:823]: Let's break this line into many final declarations.

File: lib/components/upload_form.dart [hunk 7]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+                        } else {
+                          progressText = 'Preparing...';
+                        }
+                      } else {
+                        progressText =
+                            'Your upload is in progress, but for large files the progress it not available. Please wait...';
```
Comment (matibat) [inline lib/components/upload_form.dart:829]: Has to be localized.

File: lib/components/upload_form.dart [hunk 8]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+                        children: [
+                          if (task.content != null)
+                            for (var file in task.content!)
+                              ListTile(
+                                leading: file is ARFSFileUploadMetadata
+                                    ? getIconForContentType(
+                                        file.dataContentType,
+                                        size: 24,
+                                      )
+                                    : file is ARFSFolderUploadMetatadata
```
Comment (matibat) [inline lib/components/upload_form.dart:843]: Could file be folder?

File: lib/components/upload_form.dart [hunk 9]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+                  ),
+                ),
+              ),
+            ),
+          ),
+          const SizedBox(
+            height: 8,
+          ),
+          Text(
+            'Total uploaded: ${filesize(state.progress.totalUploaded)} of ${filesize(state.progress.totalSize)}',
```
Comment (matibat) [inline lib/components/upload_form.dart:992]: Has to be localized.

File: lib/components/upload_form.dart [hunk 10]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+            style: ArDriveTypography.body
+                .buttonNormalBold(
+                    color: ArDriveTheme.of(context)
+                        .themeData
+                        .colors
+                        .themeFgDefault)
+                .copyWith(fontWeight: FontWeight.bold),
+          ),
+          Text(
+            'Files uploaded: ${state.progress.tasksContentCompleted()} of ${state.progress.tasksContentLength()}',
```
Comment (matibat) [inline lib/components/upload_form.dart:1002]: Has to be localized.

File: lib/components/upload_form.dart [hunk 11]
```diff
@@ -748,4 +762,260 @@ class _UploadFormState extends State<UploadForm> {
...
+            style: ArDriveTypography.body
+                .buttonNormalBold(
+                    color: ArDriveTheme.of(context)
+                        .themeData
+                        .colors
+                        .themeFgDefault)
+                .copyWith(fontWeight: FontWeight.bold),
+          ),
+          Text(
+            'Upload speed: ${filesize(state.progress.calculateUploadSpeed().toInt())}/s',
```
Comment (matibat) [inline lib/components/upload_form.dart:1012]: Has to be localized.

File: lib/core/crypto/crypto.dart [hunk 12]
```diff
@@ -82,37 +88,67 @@ class ArDriveCrypto {
...
+  /// Decrypts the provided transaction details and data into JSON using the provided key.
   ///
   /// Throws a [TransactionDecryptionException] if decryption fails.
-  Future<Uint8List> decryptTransactionData(
+  Future<Uint8List> decryptDataFromTransaction(
     TransactionCommonMixin transaction,
     Uint8List data,
     SecretKey key,
   ) async {
     final cipher = transaction.getTag(EntityTag.cipher);
```
Comment (matibat) [inline lib/core/crypto/crypto.dart:141]: Consider renaming it to `cipherTag`.

File: lib/core/crypto/crypto.dart [hunk 13]
```diff
@@ -82,37 +88,67 @@ class ArDriveCrypto {
...
-            .then((res) => Uint8List.fromList(res));
-      }
-    } on SecretBoxAuthenticationError catch (_) {
+    if (cipher == null || cipherIvTag == null) {
       throw TransactionDecryptionException();
     }
 
-    throw ArgumentError();
+    final decryptedData =
+        await decryptTransactionData(cipher, cipherIvTag, data, key);
```
Comment (matibat) [inline lib/core/crypto/crypto.dart:149]: Does the method take the tags, or the value of the tags?

File: lib/core/upload/upload_metadata.dart
Comment (matibat) [inline lib/core/upload/upload_metadata.dart:1]: Commented code.

File: lib/download/ardrive_downloader.dart [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    String? cipherIvString,
+  }) async* {
+    final streamDownloadResponse = await arweave.download(
+      txId: dataTx,
+      onProgress: (progress, speed) => logger.d(progress.toString()),
+    );
+
+    final streamDownload = streamDownloadResponse.$1;
+
+    Stream<Uint8List> saveStream;
```
Comment (matibat) [inline lib/download/ardrive_downloader.dart:66]: Make it late.

File: lib/download/ardrive_downloader.dart [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+      if (cipher == Cipher.aes256ctr) {
+        saveStream = await decryptTransactionDataStream(
+          cipher,
+          cipherIv,
+          streamDownload.transform(transformer),
+          keyData,
+          fileSize,
+        );
+      } else if (cipher == Cipher.aes256gcm) {
+        List<int> bytes = [];
```
Comment (matibat) [inline lib/download/ardrive_downloader.dart:82]:
```suggestion
        Uint8List bytes = Uint8List.fromList([]);
```

File: lib/download/ardrive_downloader.dart [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+        final progress = saveStatus.bytesSaved / saveStatus.totalBytes;
+
+        yield progress * 100;
+
+        // final progressPercentInt = (progress * 100).round();
+        // emit(FileDownloadWithProgress(
+        //   fileName: _file.name,
+        //   progress: progressPercentInt,
+        //   fileSize: saveStatus.totalBytes,
+        // ));
```
Comment (matibat) [inline lib/download/ardrive_downloader.dart:138]: Commented code.

File: lib/download/ardrive_downloader.dart [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    if (_cancelWithReason.isCompleted) {
+      throw Exception('Download cancelled: ${await _cancelWithReason.future}');
+    }
+    if (saveResult != true) throw Exception('Failed to save file');
+  }
+}
+
+final StreamTransformer<List<int>, Uint8List> transformer =
```
Comment (matibat) [inline lib/download/ardrive_downloader.dart:151]: Becomes unnecessary after using the Uint8List type. Can be removed.

File: lib/entities/file_entity.dart [hunk 18]
```diff
@@ -91,6 +93,8 @@ class FileEntity extends EntityWithCustomMetadata {
           data,
           fileKey,
         );
+
+        logger.d('entityJson: $entityJson');
```
Comment (matibat) [inline lib/entities/file_entity.dart:97]: Let's make it more descriptive or directly remove it.

File: lib/main.dart [hunk 19]
```diff
@@ -207,8 +207,8 @@ class AppState extends State<App> {
             communityOracle: CommunityOracle(
               ArDriveContractOracle([
                 ContractOracle(VertoContractReader()),
-                ContractOracle(RedstoneContractReader()),
-                ContractOracle(SmartweaveContractReader()),
+                // ContractOracle(RedstoneContractReader()),
+                // ContractOracle(SmartweaveContractReader()),
```
Comment (matibat) [inline lib/main.dart:211]: Commented code.

File: lib/pages/app_route_information_parser.dart [hunk 20]
```diff
@@ -11,7 +11,9 @@ class AppRouteInformationParser extends RouteInformationParser<AppRoutePath> {
   @override
   Future<AppRoutePath> parseRouteInformation(
       RouteInformation routeInformation) async {
-    final uri = Uri.parse(routeInformation.location!);
+    // TODO: Remove deprecated member use
+    // ignore: deprecated_member_use
```
Comment (matibat) [inline lib/pages/app_route_information_parser.dart:15]: Bump

File: lib/pages/drive_detail/components/drive_explorer_item_tile.dart [hunk 21]
```diff
@@ -110,43 +110,41 @@ class DriveExplorerItemTileLeading extends StatelessWidget {
...
-    } else if (FileTypeHelper.isCode(contentType)) {
-      return ArDriveIcons.fileOutlined(
-        size: size,
-      );
-    } else {
-      return ArDriveIcons.fileOutlined(
-        size: size,
-      );
-    }
+ArDriveIcon getIconForContentType(String contentType, {double size = 18}) {
```
Comment (matibat) [inline lib/pages/drive_detail/components/drive_explorer_item_tile.dart:115]: Are we re-using it? If so, then let's move the method to its own source file.

File: lib/utils/logger/logger.dart
Comment (matibat) [inline lib/utils/logger/logger.dart:1]: Unintended change?

File: packages/arconnect/LICENSE
Comment (matibat) [inline packages/arconnect/LICENSE:1]: Do we need a license for these now? Are they considered to share the same license there's in the root of the project? I don't know about legal stuff.

Comment (matibat) [reply packages/arconnect/LICENSE:1]: Same question for the rest of packages.

File: packages/arconnect/lib/src/safe_arconnect_action.dart
Comment (matibat) [inline packages/arconnect/lib/src/safe_arconnect_action.dart:1]: Commented code.

File: packages/ardrive_crypto/lib/src/authenticate.dart
Comment (matibat) [inline packages/ardrive_crypto/lib/src/authenticate.dart:1]: Remove if unnecessary.

File: packages/ardrive_crypto/lib/src/constants.dart [hunk 22]
```diff
@@ -0,0 +1,10 @@
+class Cipher {
+  static const aes256gcm = 'AES256-GCM';
+  static const aes256ctr = 'AES256-CTR';
+}
+
+class ContentType {
```
Comment (matibat) [inline packages/ardrive_crypto/lib/src/constants.dart:6]: Does it belong here?

File: packages/ardrive_crypto/lib/src/stream_aes.dart
Comment (matibat) [inline packages/ardrive_crypto/lib/src/stream_aes.dart:1]: Commented code.

File: pubspec.yaml [hunk 23]
```diff
@@ -40,16 +40,21 @@ dependencies:
...
+  arconnect:
+    path: ./packages/arconnect
+  ardrive_crypto:
+    path: ./packages/ardrive_crypto
   artemis: ^7.0.0-beta.13
   arweave:
     git:
       url: https://github.com/ardriveapp/arweave-dart.git
-      ref: v3.7.0
+      ref: PE-3697
```
Comment (matibat) [inline pubspec.yaml:55]: IMPORTANT - We gotta bump this version before releasing

Comment (matibat) [reply pubspec.yaml:55]: Same for the rest.

File: test/blocs/personal_file_download_cubit_test.dart
Comment (matibat) [inline test/blocs/personal_file_download_cubit_test.dart:1]: Let's mark it as skipped and create a tech debt ticket

File: test/core/upload/metadata_generator_test.dart
Comment (matibat) [inline test/core/upload/metadata_generator_test.dart:1]: Same comment.

File: test/utils/html_utils/tab_visibility_singleton_test.dart [hunk 24]
```diff
@@ -1,6 +1,7 @@
-import 'package:ardrive/utils/html/html_util.dart';
+import 'package:ardrive_utils/ardrive_utils.dart';
 import 'package:flutter_test/flutter_test.dart';
 
+// TODO: move this test for TabVisibilitySingleton to ardrive_utils
```
Comment (matibat) [inline test/utils/html_utils/tab_visibility_singleton_test.dart:4]: Bump.

File: lib/core/download_service.dart [hunk 25]
```diff
@@ -28,4 +31,12 @@ class _DownloadService implements DownloadService {
 
     throw Exception('Download failed');
   }
+
+  @override
+  Future<Stream<List<int>>> downloadStream(
+      String fileTxId, bool isManifest) async {
```
Comment (matibat) [inline lib/core/download_service.dart:37]: Nit: the `isManifest` parameter is being unused.

File: lib/core/upload/upload_metadata.dart [hunk 26]
```diff
@@ -56,10 +56,24 @@ class ARFSFileUploadMetadata extends ARFSUploadMetadata {
...
 
+  // without dataTxId
   @override
-  Map<String, dynamic> toJson() => _$ARFSFileUploadMetadataToJson(this);
+  Map<String, dynamic> toJson() => {
+        'name': name,
+        'size': size,
+        'lastModifiedDate': lastModifiedDate.millisecondsSinceEpoch,
+        'dataContentType': dataContentType,
+      };
```
Comment (matibat) [inline lib/core/upload/upload_metadata.dart:66]: Let's revert this change to be consistent with the rest of the classes. We can instead either:
- make another method whose name explains what's being returned; or
- make the caller remove the key it doesn't want from the result.

File: lib/blocs/upload/upload_cubit.dart [hunk 27]
```diff
@@ -471,67 +491,421 @@ class UploadCubit extends Cubit<UploadState> {
...
+        final newCompletedTasks = progress.task.where(
+          (element) =>
+              element.status == UploadStatus.complete &&
+              !completedTasks.contains(element),
+        );
+
+        for (var element in newCompletedTasks) {
+          completedTasks.add(element);
+          // TODO: Save as the file is finished the upload
+          // _saveEntityOnDB(element);
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/blocs/upload/upload_cubit.dart:779]: We got rid of this in the other PR.
